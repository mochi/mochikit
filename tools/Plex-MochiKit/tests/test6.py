import Test

import os
import sys
import Plex
from Plex import *

letter = Range("AZaz") | Any("_")
digit = Range("09")
space = Any(" \t\n")

ident = letter + Rep(letter | digit)
resword = Str("program", "unit", "uses", "const", "type", "var",
              "if", "then", "else", "while", "do", "repeat", "until",
              "for", "to", "downto", "and", "or", "not",
              "array", "of", "record", "object")
number = Rep1(digit)
string = Str("'") + (Rep(AnyBut("'")) | Str("''")) + Str("'")
diphthong = Str(":=", "<=", ">=", "<>", "..")
punct = Any("^&*()-+=[]|;:<>,./")
spaces = Rep1(space)
comment_begin = Str("{")
comment_char = AnyBut("}")
comment_end = Str("}")

lex = Lexicon([
  (resword, TEXT),
  (ident, 'ident'),
  (number, 'num'),
  (string, 'str'),
  (punct | diphthong, TEXT),
  (spaces, IGNORE),
  (comment_begin, Begin('comment')),
  State('comment', [
    (comment_char, IGNORE),
    (comment_end, Begin(''))
  ])
], 
debug = Test.debug,
timings = sys.stderr
)

Test.run(lex, "test6", debug = 0, trace = 0)


