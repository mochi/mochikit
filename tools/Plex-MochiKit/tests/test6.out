test6.in   4   0 program    'program'
test6.in   4   8 ident      'spam'
test6.in   4  12 (          '('
test6.in   4  13 ident      'input'
test6.in   4  18 ,          ','
test6.in   4  20 ident      'output'
test6.in   4  26 )          ')'
test6.in   4  27 ;          ';'
test6.in   5   0 var        'var'
test6.in   6   2 ident      'order'
test6.in   6   7 :          ':'
test6.in   6   9 ident      'integer'
test6.in   6  16 ;          ';'
test6.in   7   0 ident      'begin'
test6.in   8   2 ident      'write'
test6.in   8   7 (          '('
test6.in   8   8 str        "'How many slices of spam would you like with your eggs, sir? '"
test6.in   8  70 )          ')'
test6.in   8  71 ;          ';'
test6.in   9   2 ident      'readln'
test6.in   9   8 (          '('
test6.in   9   9 ident      'order'
test6.in   9  14 )          ')'
test6.in   9  15 ;          ';'
test6.in  10   2 if         'if'
test6.in  10   5 ident      'order'
test6.in  10  11 >=         '>='
test6.in  10  14 num        '1'
test6.in  10  16 then       'then'
test6.in  11   4 ident      'writeln'
test6.in  11  11 (          '('
test6.in  11  12 str        "'Certainly, sir.'"
test6.in  11  29 )          ')'
test6.in  12   2 else       'else'
test6.in  13   4 ident      'writeln'
test6.in  13  11 (          '('
test6.in  13  12 str        "'Sorry, sir, invalid order.'"
test6.in  13  40 )          ')'
test6.in  14   2 ident      'end'
test6.in  15   0 ident      'end'
test6.in  15   3 .          '.'
test6.in  17   0 None       ''
